### PR TITLE
Apply right name to `Route` when created from methods

### LINKED ISSUE #1552: Route naming introspection always return "method" for method endpoints
Discussion was done at https://gitter.im/encode/community. Bug was confirmed by @Kludex.

## Description
methods don't get detected on `is_function`, then we assume that `<object>.__class__.__name__ ` will give the right name (my_method on the example below) for it, but it actually gets the "method" name, which is wrong.

Unexpected behaviour seem to originate from: https://github.com/encode/starlette/blob/e086fc2da361767b532cf690e5203619bbae98aa/starlette/routing.py#L87

## Minimal example
```python
from starlette.responses import JSONResponse
from starlette.routing import Route


async def my_function(request):
    return JSONResponse({'endpoint_type': 'function'})


class MyClass:
    def __call__(self, request):
        return JSONResponse({'endpoint_type': 'class'})


class MySpecialEndpointObject:
    async def my_method(self, request):
        return JSONResponse({'endpoint_type': 'method'})


endpoint_obj = MySpecialEndpointObject()
function_route = Route('/functionEndpoint', my_function)
class_route = Route('/classEndpoint', MyClass())
method_route = Route('/methodEndpoint', endpoint_obj.my_method)

assert function_route.name == "my_function"
assert class_route.name == "MyClass"
assert method_route.name == "my_method"  # AssertionError
```

## Actual behavior

Value of `method_route.name` is `"method"`.

## Expected behavior

Value of `method_route.name` is `"my_method"`.

It could also be `"MySpecialEndpointObject_my_method"`. Reason here is to prevent ambiguity.

### PR DESCRIPTION
**Pull request by flxdot**
Fixes #1552.

Adapted existing code to fix wrong name of a Route when set automatically for routes with methods as endpoints.

### CODE REVIEW

**Comment by Kludex** [comment]
Are there other cases we are forgetting?

**Comment by flxdot** [comment]
> Are there other cases we are forgetting?

The only `Callable`s I'm aware of are:

* functions
* classes implementing `__call__`
* methods (incl. static & class methods)
* lambda

But in the nature of the lambda function is anonymous, we can not read a name.

After consulting the docs of the `inspect` module again we could simplify the implementation by using `inspect.routine`:

```python
def get_name(endpoint: typing.Callable) -> str:
    if inspect.isroutine(endpoint):
        return endpoint.__name__
    return endpoint.__class__.__name__
```

This might create issues with the lambda functions though:


```python
from starlette.responses import JSONResponse
from starlette.routing import Route


async def my_function(request):
    return JSONResponse({'endpoint_type': 'function'})


class MyClass:
    def __call__(self, request):
        return JSONResponse({'endpoint_type': 'class'})


class MySpecialEndpointObject:
    async def my_method(self, request):
        return JSONResponse({'endpoint_type': 'method'})

    @classmethod
    async def my_classmethod(self, request):
        return JSONResponse({'endpoint_type': 'classmethod'})

    @staticmethod
    async def my_staticmethod(self, request):
        return JSONResponse({'endpoint_type': 'staticmethod'})


function_route = Route('/functionEndpoint', my_function)
class_route = Route('/classEndpoint', MyClass())
endpoint_obj = MySpecialEndpointObject()
method_route = Route('/methodEndpoint', endpoint_obj.my_method)
classmethod_route = Route('/classMethodEndpoint', endpoint_obj.my_classmethod)
staticmethod_route = Route('/staticMethodEndpoint', endpoint_obj.my_staticmethod)
lambda_route = Route('/lambdaEndpoint', lambda request: JSONResponse({'endpoint_type': 'lambda'}))

assert function_route.name == "my_function"
assert class_route.name == "MyClass"
assert method_route.name == "my_method"
assert classmethod_route.name == "my_classmethod"
assert staticmethod_route.name == "my_staticmethod"
assert lambda_route.name == "lambda" # AssertionError, because it will be: "<lambda>"
```

**Comment by adriangb** [comment]
I think it's okay if lambdas aren't handled correctly

**Comment by Kludex** [comment]
Can we add a test for it?

File: starlette/routing.py [hunk 1]
```diff
@@ -84,7 +84,7 @@ async def app(scope: Scope, receive: Receive, send: Send) -> None:
 
 
 def get_name(endpoint: typing.Callable) -> str:
-    if inspect.isfunction(endpoint) or inspect.isclass(endpoint):
+    if inspect.isroutine(endpoint) or inspect.isclass(endpoint):
```
Comment (Kludex) [inline starlette/routing.py:87]: Does the behavior of lambda changes now?

**Comment by flxdot** [comment]
> Can we add a test for it?

I'm currently quite busy. But I'll add tests next week.

File: starlette/routing.py (hunk 1 above)
Comment (flxdot) [reply starlette/routing.py:87]: No:

```python3
import inspect
my_lambda = lambda: 1
assert inspect.isfunction(my_lambda) == inspect.isroutine(my_lambda)
```

**Comment by Kludex** [comment]
@flxdot Are you be able to work on this soon or can I take it over?

**Comment by flxdot** [comment]
> @flxdot Are you be able to work on this soon or can I take it over?

Friday afternoons seem to be only free time atm.

Since `get_name()` is rather an internal function I decided not to test the function, but to rather test the behaviour of the common public usage (building of routes). If you would rather like to test the function I can quickly change it.

File: tests/test_routing.py [hunk 2]
```diff
@@ -710,3 +719,54 @@ def test_duplicated_param_names():
...
+        pytest.param(
+            lambda request: JSONResponse({"endpoint_type": "lambda"}),
+            "<lambda>",
+            id="lambda",
+        ),
+    ],
+)
+def test_route_name_automatic(endpoint: typing.Callable, expected_name: str):
+
+    # Path does not matter here, as we only care about how the route name is created.
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
```

File: tests/test_routing.py [hunk 3]
```diff
@@ -710,3 +719,54 @@ def test_duplicated_param_names():
...
+    assert Route(path="/", endpoint=endpoint).name == expected_name
+
+
+def test_route_name_manual():
+
+    name = "my_custom_endpoint"
+    assert Route(path="/", endpoint=func_homepage, name=name).name == name
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
```

File: tests/test_routing.py [hunk 4]
```diff
@@ -710,3 +719,54 @@ def test_duplicated_param_names():
...
+        ),
+        pytest.param(EndpointObject, "EndpointObject", id="object"),
+        pytest.param(
+            lambda request: JSONResponse({"endpoint_type": "lambda"}),
+            "<lambda>",
+            id="lambda",
+        ),
+    ],
+)
+def test_route_name_automatic(endpoint: typing.Callable, expected_name: str):
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
def test_route_name(endpoint: typing.Callable, expected_name: str):
```

File: tests/test_routing.py [hunk 5]
```diff
@@ -710,3 +719,54 @@ def test_duplicated_param_names():
...
+        return JSONResponse({"endpoint_type": "classmethod"})
+
+    @staticmethod
+    async def my_staticmethod(self, request):
+        return JSONResponse({"endpoint_type": "staticmethod"})
+
+
+class EndpointObject:
+    def __call__(self, request):
+        return JSONResponse({"endpoint_type": "class"})
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
class Endpoint:
    async def my_method(self, request):
        ...

    @classmethod
    async def my_classmethod(cls, request):
        ...

    @staticmethod
    async def my_staticmethod(request):
        ...

    def __call__(self, request):
        ...
```

File: tests/test_routing.py [hunk 6]
```diff
@@ -710,3 +719,54 @@ def test_duplicated_param_names():
...
+            EndpointCollectionObject.my_staticmethod,
+            "my_staticmethod",
+            id="staticmethod",
+        ),
+        pytest.param(EndpointObject, "EndpointObject", id="object"),
+        pytest.param(
+            lambda request: JSONResponse({"endpoint_type": "lambda"}),
+            "<lambda>",
+            id="lambda",
+        ),
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
        pytest.param(func_homepage, "func_homepage", id="function"),
        pytest.param(Endpoint().my_method, "my_method", id="method"),
        pytest.param(
            Endpoint.my_classmethod, "my_classmethod", id="classmethod"
        ),
        pytest.param(
            Endpoint.my_staticmethod, "my_staticmethod", id="staticmethod"
        ),
        pytest.param(Endpoint, "Endpoint", id="object"),
        pytest.param(lambda request: ..., "<lambda>", id="lambda"),
```

File: tests/test_routing.py [hunk 7]
```diff
@@ -1,11 +1,20 @@
...
 import uuid
+import typing
 
 import pytest
 
 from starlette.applications import Starlette
 from starlette.responses import JSONResponse, PlainTextResponse, Response
-from starlette.routing import Host, Mount, NoMatchFound, Route, Router, WebSocketRoute
+from starlette.routing import (
+    get_name,
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
```

**Review by Kludex — CHANGES_REQUESTED** [review]

**Comment by Kludex** [comment]
Thanks for the tests!

Yeah, no need to test specifically `get_name`. I've added some comments to simplify the tests added, jfyk.

**Comment by flxdot** [comment]
Good call reducing the clutter in the test. Thanks for the hint!

File: tests/test_routing.py [hunk 8]
```diff
@@ -710,3 +712,39 @@ def test_duplicated_param_names():
...
+class Endpoint:
+    async def my_method(self, request):
+        ...
+
+    @classmethod
+    async def my_classmethod(cls, request):
+        ...
+
+    @staticmethod
+    async def my_staticmethod(self, request):
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
    async def my_staticmethod(request):
```

File: tests/test_routing.py [hunk 9]
```diff
@@ -710,3 +712,39 @@ def test_duplicated_param_names():
...
+        ),
+        pytest.param(
+            Endpoint.my_staticmethod, "my_staticmethod", id="staticmethod",
+        ),
+        pytest.param(Endpoint(), "Endpoint", id="object"),
+        pytest.param(lambda request: ..., "<lambda>", id="lambda"),
+    ],
+)
+def test_route_name(endpoint: typing.Callable, expected_name: str):
+
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
```

File: tests/test_routing.py [hunk 10]
```diff
@@ -1,11 +1,13 @@
 import functools
 import uuid
+import typing
 
 import pytest
 
 from starlette.applications import Starlette
 from starlette.responses import JSONResponse, PlainTextResponse, Response
 from starlette.routing import Host, Mount, NoMatchFound, Route, Router, WebSocketRoute
+
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
```

**Review by Kludex — CHANGES_REQUESTED** [review]

File: tests/test_routing.py [hunk 11]
```diff
@@ -1,5 +1,6 @@
 import functools
 import uuid
+import typing
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
import typing
import uuid
```

**Comment by Kludex** [comment]
Once the linter issues are resolved, we can merge it.

File: tests/test_routing.py [hunk 12]
```diff
@@ -710,3 +711,38 @@ def test_duplicated_param_names():
...
+    @classmethod
+    async def my_classmethod(cls, request):
+        ...
+
+    @staticmethod
+    async def my_staticmethod(request):
+        ...
+
+    def __call__(self, request):
+        ...
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
        ...  # pragma: no cover
```

File: tests/test_routing.py [hunk 13]
```diff
@@ -710,3 +711,38 @@ def test_duplicated_param_names():
...
+    async def my_method(self, request):
+        ...
+
+    @classmethod
+    async def my_classmethod(cls, request):
+        ...
+
+    @staticmethod
+    async def my_staticmethod(request):
+        ...
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
        ...  # pragma: no cover
```

File: tests/test_routing.py [hunk 14]
```diff
@@ -710,3 +711,38 @@ def test_duplicated_param_names():
...
         Route("/{id}/{name}/{id}/{name}", user)
+
+
+class Endpoint:
+    async def my_method(self, request):
+        ...
+
+    @classmethod
+    async def my_classmethod(cls, request):
+        ...
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
        ...  # pragma: no cover
```

File: tests/test_routing.py [hunk 15]
```diff
@@ -710,3 +711,38 @@ def test_duplicated_param_names():
         match="Duplicated param names id, name at path /{id}/{name}/{id}/{name}",
     ):
         Route("/{id}/{name}/{id}/{name}", user)
+
+
+class Endpoint:
+    async def my_method(self, request):
+        ...
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
        ...  # pragma: no cover
```

**Comment by flxdot** [comment]
sorry for all that back and forth. Should have dug deeper into the GitHub action to see, what is run.

**Review by Kludex — APPROVED** [review]

**Comment by Kludex** [comment]
Thanks @flxdot ! :)

**Pull request merged**